### PR TITLE
add SHT2x sensor type

### PR DESCRIPTION
**Pull request by hmueller01**
add SHT2xSensor class (SHT20, SHT21, SHT25)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,10 +6,11 @@ Repository for Sensirion humidity and temperature sensor support on Arduino
 - SHTC3
 - SHTW1
 - SHTW2
+- SHT2x (SHT20, SHT21, SHT25)
 - SHT3x-DIS (I2C)
 - SHT3x-ARP (ratiometric analog voltage output)
-- SHT85
 - SHT4x
+- SHT85
```
Comment (winkj) [inline README.md:12]: The SHT85 is SHT3x based, so should not be part of this commit

File: SHTSensor.cpp [hunk 2]
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
...
+{
+public:
+  SHT2xSensor(TwoWire &wire)
+      // clock stretching disabled
+      : SHTI2cSensor(0x40,     // i2cAddress
+                     0xF3F5,   // i2cCommand Hi: T, Lo: RH
+                     66,       // duration
+                     -46.85,   // a (sht_t_poly1)
+                     175.72,   // b (sht_t_poly2)
+                     65536.0,  // c (sht_t_poly3)
```
Comment (winkj) [inline SHTSensor.cpp:170]: this should likely be 65535

File: README.md (hunk 1 above)
Comment (hmueller01) [reply README.md:12]: Ah, ok. I just ordered it by number ... will revert this tomorrow.

File: SHTSensor.cpp [hunk 3]
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
...
+    uint8_t cmd[mCmd_Size];
+
+    // Upper byte is T for SHT2x Sensors
+    cmd[0] = mI2cCommand >> 8;
+    // Lower byte is RH for SHT2x Sensors
+    cmd[1] = mI2cCommand & 0xff;
+
+    // read T from SHT2x Sensor
+    if (!readFromI2c(mWire, mI2cAddress, cmd, mCmd_Size, data,
+                     EXPECTED_DATA_SIZE / 2, mDuration)) {
```
Comment (winkj) [inline SHTSensor.cpp:195]: I feel using `EXPECTED_DATA_SIZE` and later `&data[3]` is a bit confusing, but don't see a better way without changing the overall code.

Could you add a strategic comment explaining how the SHT2x is different in its readout, and that we're therefore splitting the `data` into two chunks, and therefore also halving `EXPECTED_DATA_SIZE`

File: SHTSensor.h [hunk 4]
```diff
@@ -32,6 +32,18 @@
 #include <inttypes.h>
 #include <Wire.h>
 
+//#define DEBUG_SHT_SENSOR
```
Comment (winkj) [inline SHTSensor.h:35]: Note: This could be a separate commit, but it doesn't hurt :). We could likely use this in `bool SHTI2cSensor::readSample()`, too

**Comment by winkj** [comment]
Overall, this looks great, thanks for the submission. I don't have an SHT2x to test, but will ping a few colleagues. I've added a few minor notes

**Comment by hmueller01** [comment]
FYI: I tested the code in AUTO_DETECTION on SHT21 and SHT30 ...

File: SHTSensor.cpp (hunk 2 above)
Comment (hmueller01) [reply SHTSensor.cpp:170]: Do you really want me to change this? I would prefer the datasheet value to avoid "stupid" questions ( like mine ;-) ). See PR #30

Comment (winkj) [reply SHTSensor.cpp:170]: For PR #30, I answered this for the SHTC1, since the case is clearer there

File: SHTSensor.cpp [hunk 5]
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
...
+
+    // check CRC for both RH and T with a crc init value of 0
+    if (crc8(&data[0], 2, 0) != data[2] || crc8(&data[3], 2, 0) != data[5]) {
+      DEBUG_SHT("SHT2x crc8 false\n");
+      return false;
+    }
+
+    // convert to Temperature/Humidity
+    uint16_t val;
+    val = (data[0] << 8) + data[1];
```
Comment (winkj) [inline SHTSensor.cpp]: upon re-reading the datasheet, we should clear the status bits here

For reference, here's what the sample code from https://sensirion.com/products/catalog/SHT21 does:

`u16sT &= ~0x0003;           // clear bits [1..0] (status bits)`

File: SHTSensor.cpp [hunk 6]
```diff
@@ -153,6 +153,71 @@ class SHTC1Sensor : public SHTI2cSensor
...
+    mTemperature = mA + mB * (val / mC);
+
+    val = (data[3] << 8) + data[4];
```
Comment (winkj) [inline SHTSensor.cpp]: same as above:

upon re-reading the datasheet, we should clear the status bits here

For reference, here's what the sample code from https://sensirion.com/products/catalog/SHT21 does:

`u16sT &= ~0x0003;           // clear bits [1..0] (status bits)`

File: SHTSensor.cpp (hunk 2 above)
Comment (winkj) [reply SHTSensor.cpp:170]: Upon thinking about this some more, this is likely compensated by the coefficients here. So please keep 2^16 for the SHT21

**Comment by hmueller01** [comment]
@winkj Is it possible to merge this, as I'd like to see this in my production code ... Tx a lot.

**Comment by winkj** [comment]
@hmueller01 it looks like there are still four review items open; were you planning to leave this as is?

Edit: likely my fault, the review was still open; I assumed you could see it, but chances are it was only visible to me. My apologies for that

**Comment by winkj** [comment]
@hmueller01 at the very least, the code should clear the status bits; I can add the comments on the `EXPECTED_DATA_SIZE / 2`, and I'm happy to leave the debug code as is

**Comment by hmueller01** [comment]
> @hmueller01 it looks like there are still four review items open; were you planning to leave this as is?

I only see two review items and these are resolved. Or did I overlook something?

**Comment by hmueller01** [comment]
> @hmueller01 at the very least, the code should clear the status bits; I can add the comments on the `EXPECTED_DATA_SIZE / 2`, and I'm happy to leave the debug code as is

What do you mean with "status bits"? And comments on `EXPECTED_DATA_SIZE / 2`? I do not see a comment there ...

File: SHTSensor.cpp (hunk 3 above)
Comment (winkj) [reply SHTSensor.cpp:195]: @hmueller01

File: SHTSensor.cpp (hunk 5 above)
Comment (winkj) [reply SHTSensor.cpp]: @hmueller01

File: SHTSensor.cpp (hunk 6 above)
Comment (winkj) [reply SHTSensor.cpp]: @hmueller01

File: SHTSensor.h (hunk 4 above)
Comment (winkj) [reply SHTSensor.h:35]: @hmueller01 we can ignore this; for a completely independent change like adding the debug logging, I'd typically want to have this in a separate commit, but for simplicity, we can leave this one as is

**Review by winkj — CHANGES_REQUESTED** [review]
Not sure why these have not been showing up

**Comment by winkj** [comment]
@hmueller01 the items not showing up was likely my fault, since I hadn't submitted the review. my apologies. Please let me know if the above makes sense or not.

I should be getting SHT2x inbound - hopefully arriving tomorrow - and will test & merge

File: SHTSensor.h (hunk 4 above)
Comment (hmueller01) [reply SHTSensor.h:35]: I agree, but I needed this to implement the SHT2x code. Too bad that it wasn't there in the first place. If you like we can do a new PR to add the debug feature to the "old" code to help other developers to see what goes wrong if the sensor (!= SHT2x) does not respond ...

File: SHTSensor.cpp (hunk 6 above)
Comment (hmueller01) [reply SHTSensor.cpp]: Ok, did this. Tx for the finding.

File: SHTSensor.cpp (hunk 5 above)
Comment (hmueller01) [reply SHTSensor.cpp]: Ok, did this. Tx for the finding.

File: SHTSensor.cpp (hunk 3 above)
Comment (hmueller01) [reply SHTSensor.cpp:195]: added an explanation that might help to understand the code

**Review by winkj — DISMISSED** [review]
Looks great, thanks for the updates. Will tests as soon as my SHT2x board arrives and merge

**Comment by winkj** [comment]
Got my boards, and started testing: first problem:

```c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:25: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
                    ~~~~~^~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:53: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
                                                ~~~~~^~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:223:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
     val = (data[0] << 8) + data[1] & ~0x03; // get value and clear status bits [1..0]
           ~~~~~~~~~~~~~~~^~~~~~~~~
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:226:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
     val = (data[3] << 8) + data[4] & ~0x03; // get value and clear status bits [1..0]```

**Comment by winkj** [comment]
Second error (after fixing the parentheses from the previous comment)

```           In file included from c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:32:0:
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp: In member function 'virtual bool SHT2xSensor::readSample()':
c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.h:38:40: error: 'class Serial_' has no member named 'printf'; did you mean 'print'?
   #define DEBUG_SHT(f,...) do { Serial.printf(PSTR(f), ##__VA_ARGS__); } while (0)```

**Comment by winkj** [comment]
After fixing the two problems above, two separate sensor boards both fail at readFromI2c(T); will revisit at a later point in time

Issues found (see below)

**Comment by winkj** [comment]
appears to be a timing issue with my setup: increasing duration from 66 to 100 gives me a successful results (although the status bit check still fails)

Update: found the culprit (see below)

File: SHTSensor.cpp [hunk 7]
```diff
@@ -153,6 +153,82 @@ class SHTC1Sensor : public SHTI2cSensor
...
+
+    // check CRC for both RH and T with a crc init value of 0
+    if (crc8(&data[0], 2, 0) != data[2] || crc8(&data[3], 2, 0) != data[5]) {
+      DEBUG_SHT("SHT2x crc8 false\n");
+      return false;
+    }
+
+    // check status bits [1..0] (see datasheet)
+    // bit 0: not used, bit 1: measurement type (0: temperature, 1 humidity)
+    if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
```
Comment (winkj) [inline SHTSensor.cpp]: 0x20 should likely be 0x2 here

File: SHTSensor.cpp [hunk 8]
```diff
@@ -153,6 +153,82 @@ class SHTC1Sensor : public SHTI2cSensor
...
+//
+
+class SHT2xSensor : public SHTI2cSensor
+{
+public:
+  SHT2xSensor(TwoWire &wire)
+      // clock stretching disabled
+      : SHTI2cSensor(0x40,     // i2cAddress
+                     0xF3F5,   // i2cCommand Hi: T, Lo: RH
+                     66,       // duration
```
Comment (winkj) [inline SHTSensor.cpp]: 66 -> 85

66 is the typical duration, 85 the max

File: SHTSensor.h [hunk 9]
```diff
@@ -32,6 +32,18 @@
...
+
+#ifdef DEBUG_SHT_SENSOR
+#ifdef DEBUG_ESP_PORT
+#define DEBUG_SHT(f,...) do { DEBUG_ESP_PORT.printf(PSTR(f), ##__VA_ARGS__); } while (0)
+#else
+#define DEBUG_SHT(f,...) do { Serial.printf(PSTR(f), ##__VA_ARGS__); } while (0)
```
Comment (winkj) [inline SHTSensor.h]: Serial only has print() and println(), no printf: https://www.arduino.cc/reference/en/language/functions/communication/serial/

**Comment by hmueller01** [comment]
> Got my boards, and started testing: first problem:
> 
> ```
>      if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
>                     ~~~~~^~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:216:53: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
>      if ((data[1] & 0x02 != 0x00) || (data[4] & 0x02 != 0x20)) {
>                                                 ~~~~~^~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:223:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
>      val = (data[0] << 8) + data[1] & ~0x03; // get value and clear status bits [1..0]
>            ~~~~~~~~~~~~~~~^~~~~~~~~
> c:\Users\jwinkelmann\ArduinoLocal\libraries\arduino-sht\SHTSensor.cpp:226:26: warning: suggest parentheses around '+' in operand of '&' [-Wparentheses]
>      val = (data[3] << 8) + data[4] & ~0x03; // get value and clear status bits [1..0]```
> ```

Sorry, I oversaw these warnings in my last update. Will fix this.

File: SHTSensor.cpp (hunk 7 above)
Comment (hmueller01) [reply SHTSensor.cpp]: Damn it. What a bad typo. Fixed.

File: SHTSensor.cpp (hunk 8 above)
Comment (hmueller01) [reply SHTSensor.cpp]: Hm, ok. Will increase this. All my SHT21 (5 pieces) worked with 66 ... Done.

**Review by winkj — APPROVED** [review]

File: SHTSensor.h (hunk 9 above)
Comment (hmueller01) [reply SHTSensor.h]: ESP8266 Arduino supports it ...
https://github.com/esp8266/Arduino/blob/5e0ed3a38bbe9a1e203e39eb4734dc236c41e57e/cores/esp8266/Print.h#L82C17-L82C17
But makes no sense using it, if other variants do not. Wasn't aware of that. At the end of the day I do not use parameter at the moment, so just using print is fine. Will fix this.

Comment (hmueller01) [reply SHTSensor.h]: Ok, done with [fbadd08](https://github.com/Sensirion/arduino-sht/pull/29/commits/fbadd08e6c561f99e26c0298d77e253c6613530a).

**Comment by hmueller01** [comment]
@winkj Please check again if all things work as expected. Tx!

**Comment by winkj** [comment]
> @winkj Please check again if all things work as expected. Tx!

Works great now, thanks for the quick updates. I've messed up the merge, but will do a new release later this week

**Comment by winkj** [comment]
@hmueller01 I squashed this branch into a single commit - db1c4bc738d26c89a68d0cdaa59520d831ad09b4 - and merge this into master. This is included in the v1.2.3 release that I just tagged, so that should be available via Arduino IDE for their next update cycle.

Thanks a lot for the submission and patience!

**Comment by hmueller01** [comment]
Thanks a lot for integrating my code into this project and for finding all the bugs I did. Really sorry that it wasn't more perfect in the first place.

Tested the released v1.2.3 on my SHT30 and SHT21 projects and to looks good. 

But I wonder that you released it as 1.2.3 and not as 1.3.0, as it is more than a small bugfix. Ok, at the end it's just a number. ;-)

**Comment by hmueller01** [comment]
And, documentation of 1.2.3 is a bit lame ... no hint about SHT2x

> [Version 1.2.3](https://github.com/Sensirion/arduino-sht/releases/tag/v1.2.3) [Latest](https://github.com/Sensirion/arduino-sht/releases/latest)
v1.2.3
> 
> bump version to 1.2.3

Last but not least, why is it "latest" if there is a 1.2.4 already?

**Comment by winkj** [comment]
@hmueller01 thanks for testing the new release. I wanted to get the release out ASAP and thus was short on the description, but have now updated to add the new feature:

> [Version 1.2.3](https://github.com/Sensirion/arduino-sht/releases/tag/v1.2.3)
>
> Add support for SHT2x - Contributed by hmueller01 (https://github.com/hmueller01)

Let me know if you'd prefer to be credited differently

> Last but not least, why is it "latest" if there is a 1.2.4 already?

I kept 1.2.3 as "latest" while testing of 1.2.4. Marked as latest now as the changes in .4 seem to cause no side effects
